File: Android/auikit/src/main/java/io/agora/auikit/service/imp/AUiMicSeatServiceImpl.kt
package io.agora.auikit.service.imp

import android.util.Log
import io.agora.auikit.model.AUiMicSeatInfo
import io.agora.auikit.model.AUiMicSeatStatus
import io.agora.auikit.model.AUiRoomContext
import io.agora.auikit.service.IAUiMicSeatService
import io.agora.auikit.service.callback.AUiCallback
import io.agora.auikit.service.callback.AUiException
import io.agora.auikit.service.http.CommonResp
import io.agora.auikit.service.http.HttpManager
import io.agora.auikit.service.http.Utils
import io.agora.auikit.service.http.seat.SeatEnterReq
import io.agora.auikit.service.http.seat.SeatInfoReq
import io.agora.auikit.service.http.seat.SeatInterface
import io.agora.auikit.service.http.seat.SeatLeaveReq
import io.agora.auikit.service.http.seat.SeatPickReq
import io.agora.auikit.service.rtm.AUiRtmManager
import io.agora.auikit.service.rtm.AUiRtmMsgProxyDelegate
import io.agora.auikit.utils.DelegateHelper
import io.agora.auikit.utils.GsonTools
import retrofit2.Call
import retrofit2.Response

private const val kSeatAttrKey = "micSeat"
class AUiMicSeatServiceImpl(
    private val roomContext: AUiRoomContext,
    private val channelName: String,
    private val rtmManager: AUiRtmManager
) : IAUiMicSeatService, AUiRtmMsgProxyDelegate {

    private val delegateHelper = DelegateHelper<IAUiMicSeatService.AUiMicSeatRespDelegate>()

    private var micSeats = mutableMapOf<Int, AUiMicSeatInfo>()

    init {
        rtmManager.subscribeMsg(channelName, kSeatAttrKey, this)
    }

    override fun bindRespDelegate(delegate: IAUiMicSeatService.AUiMicSeatRespDelegate?) {
        delegateHelper.bindDelegate(delegate)
    }

    override fun unbindRespDelegate(delegate: IAUiMicSeatService.AUiMicSeatRespDelegate?) {
        delegateHelper.unBindDelegate(delegate)
    }

    override fun enterSeat(seatIndex: Int, callback: AUiCallback?) {
        val param = SeatEnterReq(
            channelName,
            roomContext.currentUserInfo.userId,
            roomContext.currentUserInfo.userName,
            roomContext.currentUserInfo.userAvatar,
            seatIndex)
        HttpManager.getService(SeatInterface::class.java)
            .seatEnter(param)
            .enqueue(object : retrofit2.Callback<CommonResp<Any>> {
                override fun onResponse(call: Call<CommonResp<Any>>, response: Response<CommonResp<Any>>) {
                    if (response.body()?.code == 0) {
                        callback?.onResult(null)
                    } else {
                        callback?.onResult(Utils.errorFromResponse(response))
                    }
                }
                override fun onFailure(call: Call<CommonResp<Any>>, t: Throwable) {
                    callback?.onResult(AUiException(-1, t.message))
                }
            })
    }

    override fun autoEnterSeat(callback: AUiCallback?) {
        var toIndex: Int? = null
        for ((key, value) in micSeats) {
            if (value.seatStatus == AUiMicSeatStatus.idle) {
                toIndex = key
                break
            }
        }
        if (toIndex != null) {
            enterSeat(toIndex, callback)
        } else {
            callback?.onResult(AUiException(-1, "can not find empty mic seat"))
        }
    }

    override fun leaveSeat(callback: AUiCallback?) {
        HttpManager.getService(SeatInterface::class.java)
            .seatLeave(SeatLeaveReq(channelName, context.currentUserInfo.userId))
            .enqueue(object : retrofit2.Callback<CommonResp<Any>> {
                override fun onResponse(call: Call<CommonResp<Any>>, response: Response<CommonResp<Any>>) {
                    if (response.body()?.code == 0) {
                        callback?.onResult(null)
                    } else {
                        callback?.onResult(Utils.errorFromResponse(response))
                    }
                }
                override fun onFailure(call: Call<CommonResp<Any>>, t: Throwable) {
                    callback?.onResult(AUiException(-1, t.message))
                }
            })
    }

    override fun pickSeat(seatIndex: Int, userId: String, callback: AUiCallback?) {
        HttpManager.getService(SeatInterface::class.java)
            .seatPick(SeatPickReq(channelName, userId, seatIndex))
            .enqueue(object : retrofit2.Callback<CommonResp<Any>> {
                override fun onResponse(call: Call<CommonResp<Any>>, response: Response<CommonResp<Any>>) {
                    if (response.body()?.code == 0) {
                        callback?.onResult(null)
                    } else {
                        callback?.onResult(Utils.errorFromResponse(response))
                    }
                }
                override fun onFailure(call: Call<CommonResp<Any>>, t: Throwable) {
                    callback?.onResult(AUiException(-1, t.message))
                }
            })
    }

    override fun kickSeat(seatIndex: Int, callback: AUiCallback?) {
        HttpManager.getService(SeatInterface::class.java)
            .seatKick(SeatInfoReq(channelName, context.currentUserInfo.userId, seatIndex))
            .enqueue(object : retrofit2.Callback<CommonResp<Any>> {
                override fun onResponse(call: Call<CommonResp<Any>>, response: Response<CommonResp<Any>>) {
                    if (response.body()?.code == 0) {
                        callback?.onResult(null)
                    } else {
                        callback?.onResult(Utils.errorFromResponse(response))
                    }
                }
                override fun onFailure(call: Call<CommonResp<Any>>, t: Throwable) {
                    callback?.onResult(AUiException(-1, t.message))
                }
            })
    }

    override fun muteAudioSeat(seatIndex: Int, isMute: Boolean, callback: AUiCallback?) {
        val param = SeatInfoReq(channelName, context.currentUserInfo.userId, seatIndex)
        val service = HttpManager.getService(SeatInterface::class.java)
        val req = if (isMute) {
            service.seatAudioMute(param)
        } else {
            service.seatAudioUnMute(param)
        }
        req.enqueue(object : retrofit2.Callback<CommonResp<Any>> {
            override fun onResponse(call: Call<CommonResp<Any>>, response: Response<CommonResp<Any>>) {
                if (response.body()?.code == 0) {
                    callback?.onResult(null)
                } else {
                    callback?.onResult(Utils.errorFromResponse(response))
                }
            }
            override fun onFailure(call: Call<CommonResp<Any>>, t: Throwable) {
                callback?.onResult(AUiException(-1, t.message))
            }
        })
    }

    override fun muteVideoSeat(seatIndex: Int, isMute: Boolean, callback: AUiCallback?) {
        val param = SeatInfoReq(channelName, context.currentUserInfo.userId, seatIndex)
        val service = HttpManager.getService(SeatInterface::class.java)
        val req = if (isMute) {
            service.seatVideoMute(param)
        } else {
            service.seatVideoUnMute(param)
        }
        req.enqueue(object : retrofit2.Callback<CommonResp<Any>> {
            override fun onResponse(call: Call<CommonResp<Any>>, response: Response<CommonResp<Any>>) {
                if (response.body()?.code == 0) {
                    callback?.onResult(null)
                } else {
                    callback?.onResult(Utils.errorFromResponse(response))
                }
            }
            override fun onFailure(call: Call<CommonResp<Any>>, t: Throwable) {
                callback?.onResult(AUiException(-1, t.message))
            }
        })
    }

    override fun closeSeat(seatIndex: Int, isClose: Boolean, callback: AUiCallback?) {
        val param = SeatInfoReq(channelName, context.currentUserInfo.userId, seatIndex)
        val service = HttpManager.getService(SeatInterface::class.java)
        val req = if (isClose) {
            service.seatLock(param)
        } else {
            service.seatUnLock(param)
        }
        req.enqueue(object : retrofit2.Callback<CommonResp<Any>> {
            override fun onResponse(call: Call<CommonResp<Any>>, response: Response<CommonResp<Any>>) {
                if (response.body()?.code == 0) {
                    callback?.onResult(null)
                } else {
                    callback?.onResult(Utils.errorFromResponse(response))
                }
            }
            override fun onFailure(call: Call<CommonResp<Any>>, t: Throwable) {
                callback?.onResult(AUiException(-1, t.message))
            }
        })
    }

    override fun getMicSeatInfo(seatIndex: Int): AUiMicSeatInfo? {
        return micSeats[seatIndex]
    }
    override fun getContext() = roomContext

    override fun getChannelName() = channelName

    /** AUiRtmMsgProxyDelegate */
    override fun onMsgDidChanged(channelName: String, key: String, value: Any) {
        if (key != kSeatAttrKey) {
            return
        }
        Log.d("mic_seat_update", "class: ${value.javaClass}")
        val map: Map<String, Any> = HashMap()
        val seats = GsonTools.toBean(value as String, map.javaClass)
        Log.d("mic_seat_update", "seats: $seats")
        seats?.values?.forEach {
            val micSeat = GsonTools.toBean(GsonTools.beanToString(it), AUiMicSeatInfo::class.java) ?: return
            val index = micSeat.seatIndex
            val originMicSeat = micSeats[index]
            micSeats[index] = micSeat
            delegateHelper.notifyDelegate { delegate ->
                val newUser = micSeat.user
                if ((originMicSeat?.user == null || originMicSeat.user?.userId?.length == 0)
                    && newUser != null) {
                    Log.d("mic_seat_update", "onAnchorEnterSeat: $it")
                    delegate.onAnchorEnterSeat(index, newUser)
                }
                val originUser = originMicSeat?.user
                if ((micSeat.user == null || micSeat.user?.userId?.length == 0)
                    && originUser != null) {
                    Log.d("mic_seat_update", "onAnchorLeaveSeat: $it")
                    delegate.onAnchorLeaveSeat(index, originUser)
                }
                if ((originMicSeat?.seatStatus ?: AUiMicSeatStatus.idle) != micSeat.seatStatus) {
                    Log.d("mic_seat_update", "onSeatClose: $it")
                    delegate.onSeatClose(index, (micSeat.seatStatus == AUiMicSeatStatus.locked))
                }
                if ((originMicSeat?.muteAudio ?: 0) != micSeat.muteAudio) {
                    Log.d("mic_seat_update", "onSeatAudioMute: $it")
                    delegate.onSeatAudioMute(index, (micSeat.muteAudio != 0))
                }
                if ((originMicSeat?.muteVideo ?: 0) != micSeat.muteVideo) {
                    Log.d("mic_seat_update", "onSeatVideoMute: $it")
                    delegate.onSeatVideoMute(index, (micSeat.muteVideo != 0))
                }
            }
        }
    }
}

File: iOS/AUiKit/AUiKit/Classes/Service/Impl/AUiMicSeatServiceImpl.swift
//
//  AUiMicSeatServiceImpl.swift
//  Pods
//
//  Created by wushengtao on 2023/2/21.
//

import Foundation
import AgoraRtcKit
import YYModel


//麦位Service实现
open class AUiMicSeatServiceImpl: NSObject {
    private var respDelegates: NSHashTable<AnyObject> = NSHashTable<AnyObject>.weakObjects()
    private var channelName: String!
    private let rtmManager: AUiRtmManager!
    private let roomManager: AUiRoomManagerDelegate!
    
    private var micSeats:[Int: AUiMicSeatInfo] = [:]
    
    deinit {
        self.rtmManager.unsubscribeMsg(channelName: getChannelName(), itemKey: kSeatAttrKry, delegate: self)
        aui_info("deinit AUiMicSeatServiceImpl", tag: "AUiMicSeatServiceImpl")
    }
    
    public init(channelName: String, rtmManager: AUiRtmManager, roomManager: AUiRoomManagerDelegate) {
        self.rtmManager = rtmManager
        self.channelName = channelName
        self.roomManager = roomManager
        super.init()
        rtmManager.subscribeMsg(channelName: getChannelName(), itemKey: kSeatAttrKry, delegate: self)
        aui_info("init AUiMicSeatServiceImpl", tag: "AUiMicSeatServiceImpl")
    }
}

extension AUiMicSeatServiceImpl: AUiRtmMsgProxyDelegate {
    public func onMsgDidChanged(channelName: String, key: String, value: Any) {
        if key == kSeatAttrKry {
            aui_info("recv seat attr did changed \(value)", tag: "AUiMicSeatServiceImpl")
            guard let map = value as? [String: [String: Any]] else {return}
            map.values.forEach { element in
                guard let micSeat = AUiMicSeatInfo.yy_model(with: element) else {return}
                aui_info(" micSeat.islock \(micSeat.lockSeat) micSeat.Index = \(micSeat.seatIndex)", tag: "AUiMicSeatServiceImpl")
                let index: Int = Int(micSeat.seatIndex)
                let origMicSeat = self.micSeats[index]
//                if let origMicSeat = origMicSeat {
//                    origMicSeat.user = roomManager.getUserInfo(by: origMicSeat.userId)
//                }
//                micSeat.user = roomManager.getUserInfo(by: micSeat.userId)
                self.micSeats[index] = micSeat
                self.respDelegates.allObjects.forEach { obj in
                    guard let delegate = obj as? AUiMicSeatRespDelegate else {
                        return
                    }
                    
                    if let origUser = origMicSeat?.user, origUser.userId.count > 0, micSeat.user?.userId ?? "" != origUser.userId {
                        delegate.onAnchorLeaveSeat(seatIndex: index, user: origUser)
                    }
                    
                    if let user = micSeat.user, user.userId.count > 0, origMicSeat?.user?.userId ?? "" != user.userId {
                        delegate.onAnchorEnterSeat(seatIndex: index, user: user)
                    }
                    
                    if origMicSeat?.lockSeat ?? .idle != micSeat.lockSeat {
                        delegate.onSeatClose(seatIndex: index, isClose: micSeat.lockSeat == .locked)
                    }
                    
                    if origMicSeat?.muteAudio ?? false != micSeat.muteAudio {
                        delegate.onSeatAudioMute(seatIndex: index, isMute: micSeat.muteAudio)
                    }
                    
                    if origMicSeat?.muteVideo ?? false != micSeat.muteVideo {
                        delegate.onSeatVideoMute(seatIndex: index, isMute: micSeat.muteVideo)
                    }
                    /*
                    if origMicSeat?.muteVideo != micSeat.muteVideo {
                        delegate.onSeatVideoMute(seatIndex: index, isMute: micSeat.muteVideo)
                    }
                    
                    if origMicSeat?.muteAudio != micSeat.muteAudio {
                        delegate.onSeatAudioMute(seatIndex: index, isMute: micSeat.muteAudio)
                    }
                     */
                }
            }
        }
    }
}


extension AUiMicSeatServiceImpl: AUiMicSeatServiceDelegate {
    public func getChannelName() -> String {
        return channelName
    }
    
    public func bindRespDelegate(delegate: AUiMicSeatRespDelegate) {
        respDelegates.add(delegate)
    }
    
    public func unbindRespDelegate(delegate: AUiMicSeatRespDelegate) {
        respDelegates.remove(delegate)
    }
    
    public func enterSeat(seatIndex: Int, callback: @escaping (Error?) -> ()) {
//        if let _ = self.micSeats.values.filter({ $0.userId == self.getRoomContext().currentUserInfo.userId }).first {
//            callback(nil)
//            return
//        }
        let model = AUiSeatEnterNetworkModel()
        model.roomId = channelName
        model.userAvatar = getRoomContext().currentUserInfo.userAvatar
        model.userId = getRoomContext().currentUserInfo.userId
        model.userName = getRoomContext().currentUserInfo.userName
//        model.user = getRoomContext().currentUserInfo
        model.micSeatNo = seatIndex
        model.request { error, _ in
            callback(error)
        }

        return
        //mock
        /*
        if let _ = self.micSeats.values.filter({ $0.userId == self.getRoomContext().currentUserInfo.userId }).first {
            callback(nil)
            return
        }
        
        var seatMap: [String: [String: Any]] = [:]
        
        self.micSeats.forEach { (key: Int, value: AUiMicSeatInfo) in
            var map = value.yy_modelToJSONObject() as? [String: Any] ?? [:]
            if key == seatIndex {
                map["user"] = self.getRoomContext().currentUserInfo.yy_modelToJSONObject()
            }
            seatMap["\(key)"] = map
        }
        
        let data = try! JSONSerialization.data(withJSONObject: seatMap, options: .prettyPrinted)
        let str = String(data: data, encoding: .utf8)!
        let metaData = ["seat": str]
        self.rtmManager.setMetadata(channelName: channelName, metadata: metaData) { error in
            callback(nil)
        }
         */
    }
    
    public func leaveSeat(callback: @escaping (Error?) -> ()) {
        
        let model = AUiSeatLeaveNetworkModel()
        model.roomId = channelName
        model.userId = getRoomContext().currentUserInfo.userId
//        model.micSeatNo = seatIndex
        model.request { error, _ in
            callback(error)
        }
        return
        //mock
        /*
        guard let seat = self.micSeats.values.filter({ $0.userId == self.getRoomContext().currentUserInfo.userId }).first else {
            callback(nil)
            return
        }
        
        var seatMap: [String: [String: Any]] = [:]
        
        self.micSeats.forEach { (key: Int, value: AUiMicSeatInfo) in
            var map = value.yy_modelToJSONObject() as? [String: Any] ?? [:]
            if key == seat.seatIndex {
                map.removeValue(forKey: "user")
            }
            seatMap["\(key)"] = map
        }
        
        let str = (seatMap as AnyObject).yy_modelToJSONString() ?? ""
        let metaData = ["seat": str]
        self.rtmManager.setMetadata(channelName: channelName, metadata: metaData) { error in
            callback(nil)
        }
         */
    }
    
    public func pickSeat(seatIndex: Int, userId: String, callback: @escaping (Error?) -> ()) {
        //mock
//        guard let seat = self.micSeats[seatIndex], seat.user == nil else {
//            callback(nil)
//            return
//        }
//        
//        var seatMap: [String: [String: Any]] = [:]
//        
//        self.micSeats.forEach { (key: Int, value: AUiMicSeatInfo) in
//            var map = value.yy_modelToJSONObject() as? [String: Any] ?? [:]
//            if key == seatIndex {
//                let user = AUiUserThumbnailInfo()
//                user.userId = userId
//                user.userName = userId
//                map["user"] = user.yy_modelToJSONObject()
//            }
//            seatMap["\(key)"] = map
//        }
//        
//        let str = (seatMap as AnyObject).yy_modelToJSONString() ?? ""
//        let metaData = ["seat": str]
//        self.rtmManager.setMetadata(channelName: channelName, metadata: metaData) { error in
//            callback(nil)
//        }
    }
    
    public func kickSeat(seatIndex: Int, callback: @escaping (Error?) -> ()) {
        let model = AUiSeatkickNetworkModel()
        model.roomId = channelName
        model.userId = getRoomContext().currentUserInfo.userId
        model.micSeatNo = seatIndex
        model.request { error, _ in
            callback(error)
        }
        return
        //mock
        /*
        guard let seat = self.micSeats[seatIndex] else {
            callback(nil)
            return
        }
        
        var seatMap: [String: [String: Any]] = [:]
        
        self.micSeats.forEach { (key: Int, value: AUiMicSeatInfo) in
            var map = value.yy_modelToJSONObject() as? [String: Any] ?? [:]
            if key == seatIndex {
                map = [
                   "seatNo": seat.seatIndex,
               ]
            }
            seatMap["\(key)"] = map
        }
        
        let str = (seatMap as AnyObject).yy_modelToJSONString() ?? ""
        let metaData = ["seat": str]
        self.rtmManager.setMetadata(channelName: channelName, metadata: metaData) { error in
            callback(nil)
        }
         */
    }
    
    public func muteAudioSeat(seatIndex: Int, isMute: Bool, callback: @escaping (Error?) -> ()) {
        if isMute {
            let model = AUiSeatMuteAudioNetworkModel()
            model.roomId = channelName
            model.micSeatNo = seatIndex
            model.userId = getRoomContext().currentUserInfo.userId
            model.request { error, _ in
                callback(error)
            }
        }else {
            let model = AUiSeatUnMuteAudioNetworkModel()
            model.roomId = channelName
            model.micSeatNo = seatIndex
            model.userId = getRoomContext().currentUserInfo.userId
            model.request { error, _ in
                callback(error)
            }
        }
        //mock
        /*
        guard let _ = self.micSeats[seatIndex] else {
            //TODO: fatel error
            callback(nil)
            return
        }
        
        var seatMap: [String: [String: Any]] = [:]
        
        self.micSeats.forEach { (key: Int, value: AUiMicSeatInfo) in
            var map = value.yy_modelToJSONObject() as? [String: Any] ?? [:]
            if key == seatIndex {
                map["isMuteAudio"] = isMute
            }
            
            seatMap["\(key)"] = map
        }
        
        let data = try! JSONSerialization.data(withJSONObject: seatMap, options: .prettyPrinted)
        let str = String(data: data, encoding: .utf8)!
        let metaData = ["seat": str]
        self.rtmManager.setMetadata(channelName: channelName, metadata: metaData) { error in
            callback(nil)
        }
         */
    }
    
    public func muteVideoSeat(seatIndex: Int, isMute: Bool, callback: @escaping AUiCallback) {
        if isMute {
            let model = AUiSeatMuteVideoNetworkModel()
            model.roomId = channelName
            model.micSeatNo = seatIndex
            model.userId = getRoomContext().currentUserInfo.userId
            model.request { error, _ in
                callback(error)
            }
        }else {
            let model = AUiSeatUnMuteVideoNetworkModel()
            model.roomId = channelName
            model.micSeatNo = seatIndex
            model.userId = getRoomContext().currentUserInfo.userId
            model.request { error, _ in
                callback(error)
            }
        }
        //mock
        /*
        guard let _ = self.micSeats[seatIndex] else {
            //TODO: fatel error
            callback(nil)
            return
        }
        
        var seatMap: [String: [String: Any]] = [:]
        
        self.micSeats.forEach { (key: Int, value: AUiMicSeatInfo) in
            var map = value.yy_modelToJSONObject() as? [String: Any] ?? [:]
            if key == seatIndex {
                map["isMuteVideo"] = isMute
            }
            seatMap["\(key)"] = map
        }
        
        let data = try! JSONSerialization.data(withJSONObject: seatMap, options: .prettyPrinted)
        let str = String(data: data, encoding: .utf8)!
        let metaData = ["seat": str]
        self.rtmManager.setMetadata(channelName: channelName, metadata: metaData) { error in
            callback(nil)
        }
         */
    }
    
    public func closeSeat(seatIndex: Int, isClose: Bool, callback: @escaping (Error?) -> ()) {
        if isClose {
            let model = AUiSeatLockNetworkModel()
            model.roomId = channelName
            model.micSeatNo = seatIndex
            model.userId = getRoomContext().currentUserInfo.userId
            model.request { error, _ in
                callback(error)
            }
        }else {
            let model = AUiSeatUnLockNetworkModel()
            model.roomId = channelName
            model.micSeatNo = seatIndex
            model.userId = getRoomContext().currentUserInfo.userId
            model.request { error, _ in
                callback(error)
            }
        }
        
        return
        //mock
        /*
        guard let seat = self.micSeats[seatIndex], seat.userId == nil else {
            callback(nil)
            return
        }
        
        var seatMap: [String: [String: Any]] = [:]
        
        self.micSeats.forEach { (key: Int, value: AUiMicSeatInfo) in
            var map = value.yy_modelToJSONObject() as? [String: Any] ?? [:]
            if key == seatIndex {
                map["isLockSeat"] = isClose
            }
            seatMap["\(key)"] = map
        }
        
        let data = try! JSONSerialization.data(withJSONObject: seatMap, options: .prettyPrinted)
        let str = String(data: data, encoding: .utf8)!
        let metaData = ["seat": str]
        self.rtmManager.setMetadata(channelName: channelName, metadata: metaData) { error in
            callback(nil)
        }
         */
    }
}
